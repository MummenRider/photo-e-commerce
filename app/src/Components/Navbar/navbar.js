import React, { useState, useContext, createContext } from "react";
import {
  Container,
  Header,
  Logo,
  Menu,
  Item,
  Button,
  MenuSide,
} from "Components/Navbar/navbar.styles";
import useScroll from "Hooks/useScroll";

const ToggleMenuContext = createContext();

export default function Navbar({ isBigDevice, children, ...restProps }) {
  const { isScrolled } = useScroll(false);

  return (
    <Header
      initial={{ padding: !isScrolled && !isBigDevice ? "5vh 0" : "3vh 0" }}
      animate={{
        padding:
          isScrolled && !isBigDevice
            ? "2vh 0"
            : isScrolled && isBigDevice
            ? "1vh 0"
            : !isScrolled && !isBigDevice
            ? "5vh 0"
            : "3vh 0",
      }}
      transition={{ duration: 0.5, type: "tween" }}
      isScrolled={isScrolled}
      data-testid="nav-background"
      {...restProps}
    >
      {children}
    </Header>
  );
}

Navbar.Container = function NavbarContainer({ children, ...restProps }) {
  const [toggleMenu, setToggleMenu] = useState(ToggleMenuContext);
  return (
    <ToggleMenuContext.Provider value={{ toggleMenu, setToggleMenu }}>
      <Container {...restProps}>{children}</Container>
    </ToggleMenuContext.Provider>
  );
};

Navbar.Logo = function NavbarLogo({ children, ...restProps }) {
  return <Logo {...restProps}>{children}</Logo>;
};

Navbar.MenuSide = function NavbarMenuSide({
  isBigDevice,
  children,
  ...restProps
}) {
  return !isBigDevice ? <MenuSide {...restProps}>{children}</MenuSide> : null;
};

Navbar.Menu = function NavbarMenu({ isBigDevice, children, ...restProps }) {
  const { toggleMenu } = useContext(ToggleMenuContext);

  return isBigDevice ? (
    <Menu data-testid="nav-menu" isOpen={toggleMenu} {...restProps}>
      {children}
    </Menu>
  ) : null;
};

Navbar.Item = function NavbarItem({ children, ...restProps }) {
  return <Item {...restProps}>{children}</Item>;
};

Navbar.Button = function NavbarButton({
  setIsOpen,
  isBigDevice,
  children,
  ...restProps
}) {
  return !isBigDevice ? (
    <Button data-testid="menu-icon" onClick={setIsOpen} {...restProps}>
      {children}
    </Button>
  ) : null;
};
