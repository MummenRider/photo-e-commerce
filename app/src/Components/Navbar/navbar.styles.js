import styled from "styled-components";
import { NavLink as Link } from "react-router-dom";

export const Nav = styled.nav`
  display: flex;
  justify-content: space-between;
  width: 90%;
  padding: 3rem 0rem;
  margin: auto;
  font-family: "Montserrat", sans-serif;
  align-items: center;
  flex-wrap: wrap;
  @media (max-width: 768px) {
    width: 100%;
  }
`;

export const Logo = styled(Link)`
  color: #ffffff;
  font-size: 2rem;
  font-weight: 400;
  text-decoration: none;
  text-align: center;

  @media (max-width: 768px) {
    padding-left: 4rem;
    padding-bottom: 2rem;
  }
`;

export const Hamburger = styled.div`
  display: none;
  flex-direction: column;
  cursor: pointer;

  span {
    height: 2px;
    width: 25px;
    background: white;
    margin-bottom: 4px;
    border-radius: 5px;
  }

  @media (max-width: 768px) {
    display: flex;
    padding-right: 4rem;
    padding-bottom: 2rem;
  }
`;

export const Menu = styled.div`
  display: flex;
  justify-content: space-between;
  color: #ffffff;
  position: relative;
  width: 40rem;

  @media (max-width: 768px) {
    overflow: hidden;
    flex-direction: column;
    width: 100%;
    max-height: ${({ isOpen }) => (isOpen ? "300px" : "0px")};
    transition: max-height 300ms ease-in;
    background-color: #353535;
  }
`;

export const MenuLink = styled(Link)`
  color: #ffffff;
  font-size: 1.4rem;
  font-weight: 100;
  letter-spacing: 0.2rem;
  text-decoration: none;
  text-align: center;
  position: relative;

  &::after {
    content: "";
    position: absolute;
    border-bottom: 0.019rem solid white;
    bottom: 0;
    left: 0;
    width: 100%;
    transform-origin: right;
    transform: scaleX(0);
    transition: transform 250ms ease-in;
  }

  &:hover::after {
    transform: scaleX(1);
    transform-origin: left;
  }

  @media (max-width: 768px) {
    padding: 2rem 0rem;

    &:hover::after {
      transform: scaleX(0);
    }
  }
`;
