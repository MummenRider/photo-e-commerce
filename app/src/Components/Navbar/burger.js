import React from "react";
import { motion } from "framer-motion";
import { BurgerContainer, BurgerLine } from "./navbar.styles";

export default function Burger() {
  return (
    <BurgerContainer>
      <svg
        xmlns="http://www.w3.org/2000/svg"
        width="22.263"
        height="22.263"
        viewBox="0 0 22.263 22.263"
      >
        <g
          id="Group_127"
          data-name="Group 127"
          transform="translate(-183.869 -172.369)"
        >
          <line
            id="Line_46"
            data-name="Line 46"
            x2="20"
            y2="20"
            transform="translate(185 173.5)"
            fill="none"
            stroke="#313b41"
            stroke-linecap="round"
            stroke-width="1.6"
          />
          <line
            id="Line_47"
            data-name="Line 47"
            x2="20"
            y2="20"
            transform="translate(185 193.5) rotate(-90)"
            fill="none"
            stroke="#313b41"
            stroke-linecap="round"
            stroke-width="1.6"
          />
        </g>
      </svg>
    </BurgerContainer>
  );
}
