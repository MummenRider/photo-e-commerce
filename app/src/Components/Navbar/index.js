import React, { useState } from "react";

import { Nav, Logo, Hamburger, Menu, MenuLink } from "./navbar.styles";

const Navbar = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <header>
      <Nav>
        <Logo to="/">
          <h3>LOGO</h3>
        </Logo>
        <Hamburger onClick={() => setIsOpen(!isOpen)}>
          <span />
          <span />
          <span />
        </Hamburger>
        <Menu isOpen={isOpen}>
          <MenuLink to="/about">ABOUT</MenuLink>
          <MenuLink to="/services">SERVICES</MenuLink>
          <MenuLink to="/contact">CONTACT</MenuLink>
        </Menu>
      </Nav>
    </header>
  );
};

export default Navbar;
