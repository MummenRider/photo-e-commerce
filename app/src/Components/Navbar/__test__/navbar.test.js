import { cleanup, render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import "jest-styled-components";
import { Hamburger, Menu } from "Components/Navbar/navbar.styles.js";
import renderer from "react-test-renderer";
import Navbar from "Components/Navbar/navbar.js";

afterEach(cleanup);

const links = [
  { text: "ABOUT", location: "/about" },
  { text: "SERVICES", location: "/services" },
  { text: "CONTACT", location: "/contact" },
];

it("should render navbar component", () => {
  const { container } = render(
    <Router>
      <Navbar />
    </Router>
  );

  expect(container.firstChild).toMatchSnapshot();
});

describe("anchor tag", () => {
  it("logo is visible and is linked to home page", () => {
    render(
      <Router>
        <Navbar />
      </Router>
    );
    const logo = screen.getByTestId("logo");
    expect(screen.getByText("LOGO")).toBeVisible();
    expect(logo).toHaveAttribute("href", "/");
  });

  it.each(links)("Nav Bar have %s link.", (link) => {
    render(
      <Router>
        <Navbar />
      </Router>
    );
    const linkDom = screen.getByText(link.text);
    expect(linkDom).toHaveAttribute("href", link.location);
  });
});

describe("in mobile mode", () => {
  it("hamburger value should be hidden", () => {
    const { container } = render(<Hamburger />);
    expect(container.firstChild).toHaveStyleRule("display", "none");
  });

  it("hamburger value should be visible", () => {
    const { container } = render(<Hamburger />);
    expect(container.firstChild).toHaveStyleRule("display", "flex", {
      media: "(max-width: 768px)",
    });
  });

  it("menu links should be hidden", () => {
    const tree = renderer.create(<Menu isOpen />).toJSON();
    expect(tree).not.toHaveStyleRule("max-height", "0px", {
      media: "(max-width: 768px)",
    });
  });

  it("menu links should be visible", () => {
    const tree = renderer.create(<Menu isOpen />).toJSON();
    expect(tree).toHaveStyleRule("max-height", "300px", {
      media: "(max-width: 768px)",
    });
  });
});
