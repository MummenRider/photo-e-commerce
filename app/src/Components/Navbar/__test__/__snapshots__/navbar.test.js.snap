// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render navbar component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: fixed;
  padding: 2rem 0;
  top: 0;
  background-color: transparent;
  font-family: "Montserrat",sans-serif;
  min-height: 10vh;
  width: 100%;
  z-index: 999;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 90%;
  margin: auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  color: white;
  font-size: 2rem;
  font-weight: 400;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
}

.c3 {
  display: none;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  cursor: pointer;
}

.c3 span {
  height: 2px;
  width: 25px;
  background: white;
  margin-bottom: 4px;
  border-radius: 5px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  color: #ffffff;
  position: relative;
  width: 40rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  color: #ffffff;
  font-size: 1.8rem;
  font-weight: 100;
  -webkit-letter-spacing: 0.2rem;
  -moz-letter-spacing: 0.2rem;
  -ms-letter-spacing: 0.2rem;
  letter-spacing: 0.2rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-align: center;
  position: relative;
}

.c5::after {
  content: "";
  position: absolute;
  border-bottom: 0.019rem solid white;
  bottom: 0;
  left: 0;
  width: 100%;
  -webkit-transform-origin: right;
  -ms-transform-origin: right;
  transform-origin: right;
  -webkit-transform: scaleX(0);
  -ms-transform: scaleX(0);
  transform: scaleX(0);
  -webkit-transition: -webkit-transform 250ms ease-in;
  -webkit-transition: transform 250ms ease-in;
  transition: transform 250ms ease-in;
}

.c5:hover::after {
  -webkit-transform: scaleX(1);
  -ms-transform: scaleX(1);
  transform: scaleX(1);
  -webkit-transform-origin: left;
  -ms-transform-origin: left;
  transform-origin: left;
}

@media screen and (max-width:768px) {
  .c1 {
    padding: 1rem 0;
    width: 100%;
    margin: 0;
  }
}

@media screen and (max-width:768px) {
  .c2 {
    padding: 0 0 2rem 4rem;
  }
}

@media (max-width:768px) {
  .c3 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    padding: 0 4rem 2rem 0;
  }
}

@media (max-width:768px) {
  .c4 {
    overflow: hidden;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
    width: 100%;
    max-height: 0px;
    -webkit-transition: max-height 300ms ease-in;
    transition: max-height 300ms ease-in;
    background-color: #353535;
  }
}

@media screen and (max-width:768px) {
  .c5 {
    padding: 2rem 0rem;
  }

  .c5:hover::after {
    -webkit-transform: scaleX(0);
    -ms-transform: scaleX(0);
    transform: scaleX(0);
  }
}

<header
  class="c0"
>
  <nav
    class="c1"
    data-testid="nav"
  >
    <a
      aria-current="page"
      class="c2 active"
      data-testid="logo"
      href="/"
    >
      <h3>
        LOGO
      </h3>
    </a>
    <div
      class="c3"
    >
      <span />
      <span />
      <span />
    </div>
    <div
      class="c4"
    >
      <a
        class="c5"
        data-testid="about"
        href="/about"
      >
        ABOUT
      </a>
      <a
        class="c5"
        href="/services"
      >
        SERVICES
      </a>
      <a
        class="c5"
        href="/contact"
      >
        CONTACT
      </a>
    </div>
  </nav>
</header>
`;
